// socket.io-request/lib/index.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _error = require("./error");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

module.exports = function (io, options) {
  return new SocketIORequest(io, options);
};

var SocketIORequest = function () {
  function SocketIORequest(io) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, SocketIORequest);

    this.io = io;
    this.options = Object.assign({
      event: "socket.io-request",
      timeout: 90000
    }, options);
  }

  _createClass(SocketIORequest, [{
    key: "request",
    value: function request(method, data) {
      var _this = this;

      if (typeof method !== "string") throw new Error('argument "method" is missing');

      return new Promise(function (resolve, reject) {
        _this.io.emit(_this.options.event, { method: method, data: data }, function (res) {
          clearTimeout(timeout);
          _this.io.removeListener("disconnect", onDisconnect);
          if (res.error) return reject((0, _error.convertObjectToError)(res.error));
          resolve(res.data);
        });

        var onDisconnect = function onDisconnect() {
          clearTimeout(timeout);
          reject(new _error.SocketIOError("disconnect"));
        };

        var timeout = setTimeout(function () {
          _this.io.removeListener("disconnect", onDisconnect);
          reject(new _error.TimeoutError("exceeded " + _this.options.timeout + " (msec)"));
        }, _this.options.timeout);

        _this.io.once("disconnect", onDisconnect);
      });
    }
  }, {
    key: "response",
    value: function response(method, callback) {
      if (typeof method !== "string") throw new Error('argument "method" is missing');
      if (typeof callback !== "function") throw new Error('"callback" must be a function');
      this.io.on(this.options.event, function (req, ack) {
        if (req.method !== method) return;
        var res = function res(data) {
          ack({ data: data });
        };
        res.error = function (err) {
          ack({ error: (0, _error.convertErrorToObject)(err) });
        };
        callback(req.data, res);
      });
    }
  }]);

  return SocketIORequest;
}();
